Backend login signup

shahgrishma123 opened the pull request:
feature: added authentication using JWT.
feature: added cookies using cookies parser.
feature: added login/register route for the backend.

shahgrishma123 committed:
Added Name to .env file,Grishma Shah

shahgrishma123 committed:
create backend-login-signup.

shahgrishma123 committed:
Resolved merge conflict

johnsciutto committed:
Merged User methods created by John with Grishma's

The functions to perform CRUD operations are now available through static and
instance methods of the user model.

johnsciutto committed:
Linting with prettier, no other edits

johnsciutto committed:
Merged changes from backend-login-signup

johnsciutto committed:
Add softDeletionPlugin to userSchema

johnsciutto committed:
Change name of create user function

shahgrishma123 committed:
some bugfix

shahgrishma123 committed:
hasProprty file added

johnsciutto committed:
Updated linting in a lot of files. No big changes.

Commit previous to pulling new updates from parent branch.

jessdesr commented on server/routes/login.js:
```diff
@@ -0,0 +1,36 @@
+    //res.redirect("main");
+});
+
+router.get("/",function(req,res){
```
This route is unnecessary, it can be removed

shahgrishma123 replied:
ok. I will remove this.

jessdesr commented on server/routes/login.js:
```diff
@@ -0,0 +1,36 @@
+        companyName: req.body.companyName,
+        accessToken: token
+    });
+    //res.redirect("main");
```
Remove comments

shahgrishma123 replied:
ok. I will change this.

jessdesr commented on server/routes/register.js:
```diff
@@ -0,0 +1,84 @@
+    if(!hp.hasValidProperty(message))
+        return res.status(400).send(message);
+    
+    let result = verifyRegister.checkEmailExistedInModel(req,res,next);
```
this verifyRegister shouldn't be required, you should get an error from the DB if there is an attempt to save a user with an email that already exists

shahgrishma123 replied:
ok. I will change this.

johnsciutto committed:
After merge with parent branch.

Merge branch 'backend-login-signup' into backend-login-signup-john.

jessdesr commented on server/config/auth.config.js:
```diff
@@ -0,0 +1,5 @@
+secret = {
+    secret: "Our little secret."
```
This should be added to the environment file instead of being a separate config.

jessdesr commented on server/app.js:
```diff
@@ -17,8 +19,21 @@ app.use(urlencoded({ extended: false }));
 app.use("/", indexRouter);
-app.use("/ping", pingRouter);
+
+/*app.use(function(req, res, next) {
```
Don't need to leave this in

jessdesr commented on server/app.js:
```diff
@@ -17,8 +19,21 @@ app.use(urlencoded({ extended: false }));
 app.use(cookieParser());
 app.use(express.static(join(__dirname, "public")));
 
+db.mongoose.connect("mongodb://localhost:27017/userDB", {useNewUrlParser: true});
```
All of these DB things should be changed to use @johnsciutto's code.  We can merge that in now.

shahgrishma123 replied:
This line of code for my testing purpose. Anyway, I am going to delete that one. Also, userModel folder is unnecessary. It is only for my testing purpose. I think we can remove that one after @johnsciutto completes his curd operation. I am going to use his methods in my code.

johnsciutto replied:
Hi @shahgrishma123. I'm done with all the CRUD methods, I'm going to merge into this branch now. I added detailed instructions in the README for all the database-related methods. 
Also, because you're going to delete the "userModel" folder, I moved all the functions that are still needed to the "models" folder and to the "utilities" folder.

jessdesr commented on server/userModel/userData.js:
```diff
@@ -0,0 +1,59 @@
+    return newUser;
+}
+
+function saveDataToUserModel(newUser){
```
Most of these helpers aren't really required. You can handle the logic within the routes themselves.

jessdesr commented on server/utils/hasProperty.js:
```diff
@@ -0,0 +1,16 @@
+function hasValidProperty(name){
+    return name.hasOwnProperty('valid');
```
Same for these, the response from mongoDB should be enough to validate the results

shahgrishma123 replied:
For this change, I have to change the design. I am going to pass all my (req,res) param to controllers. It will handle all the errors with its status.

jessdesr commented:
You will need to refactor this branch once @johnsciutto's changes are merged into your branch.  Take a look at his PR to get an idea of what you'll need to change.

johnsciutto committed:
Completed mongoose configuration and functions.

- Finished and manually tested mongoose methods in online test database.
- Formatting of userData with prettier.

johnsciutto committed:
Change name of encryptedPasswordWithSalt

To verb to reflect the action that the function performs.

johnsciutto committed:
Refactor createNewUserObject

To be a one function trigger to tigger the creation, validation, hashing and
saving of a user in the database.

johnsciutto committed:
Delete unnecesary functions.

Their functionality is already covered by static and instance methods.

johnsciutto committed:
Better database error handling

johnsciutto committed:
Tightening security in generateAuthToken

johnsciutto committed:
Changed registerUser so that it works async.

Also eliminated unnecesary functions and imports.

johnsciutto committed:
Apply prettier to loginController.js

johnsciutto committed:
Refactor loginController to work asynchronously

And with the mongoose static method (findByEmail).

johnsciutto committed:
Include database functions usage

johnsciutto committed:
Add _ notation to private instance methods

johnsciutto committed:
Improve error messages in loginUser

johnsciutto committed:
Small changes and edits

- Changed `User.findByEmail(...)` to make sure it returns an object and not an
  array with a single object.
- Added @private tag to private methods.

johnsciutto committed:
Moved user model to /models/user-model.js

johnsciutto committed:
Moved encryptPasswordWithSalt to /utils/database.js

johnsciutto committed:
Moved authentication utility functions to /utils/

- Created `authentication.js` in `/utils/`.
- Moved `isPasswordValid` and `generateAuthToken`.
- Wrote documentation for both functions.

johnsciutto committed:
Merge pull request #60 from hatchways/backend-login-signup-john

Built CRUD operations on Grishma's work.

shahgrishma123 committed:
Remove comments from app.js file

shahgrishma123 committed:
Removed comments from login.js,register.js. Removed get route from login.js

shahgrishma123 committed:
Removed verifyRegister.js, changed the path on registerController.js for userModel

shahgrishma123 committed:
Removed hasProperty file from utils. changed some validation  structure. Added object into generateAuthentication function. Added Cookies file into utils.

shahgrishma123 committed:
Removed comments from user-models

shahgrishma123 committed:
Removed userModel folder from the project

shahgrishma123 committed:
update readme file

shahgrishma123 committed:
Changed Cookies_Expire and JWT_expire name in file. updated Readme file.

shahgrishma123 committed:
Added auth.js in router file and removed login and regiser route from router folder.

shahgrishma123 committed:
Moved geenrateAuthToken method to userRegister method to controller.Also throw some error from generateAuthToken

jessdesr commented on server/controllers/loginController.js:
```diff
@@ -0,0 +1,38 @@
+
+async function loginUser(req,res) {
+  const user = await User.findByEmail(req.body.email);
+  console.log(user);
```
remove this console log

jessdesr reviewed (approved):
Just delete the comment then it can be merged!

shahgrishma123 committed:
remove comment from loginController

shahgrishma123 committed:
Merge branch 'dev' into backend-login-signup
